controller/konnect/reconciler_kongplugin.go

```diff
@@ -0,0 +1,256 @@
+
+	"github.com/samber/lo"
+	k8serrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+
+	"github.com/kong/gateway-operator/controller/pkg/log"
+	"github.com/kong/gateway-operator/pkg/consts"
+	k8sutils "github.com/kong/gateway-operator/pkg/utils/kubernetes"
+
+	configurationv1 "github.com/kong/kubernetes-configuration/api/configuration/v1"
+	configurationv1alpha1 "github.com/kong/kubernetes-configuration/api/configuration/v1alpha1"
+)
+
+// KongPluginReconciler reconciles a KongPlugin object.
+type KongPluginReconciler struct {
+	DevelopmentMode bool
+	Client          client.Client
+}
+
+// NewKongPluginReconciler creates a new KongPluginReconciler.
+func NewKongPluginReconciler(
+	developmentMode bool,
+	client client.Client,
+) *KongPluginReconciler {
+	return &KongPluginReconciler{
+		DevelopmentMode: developmentMode,
+		Client:          client,
+	}
+}
+
+// SetupWithManager sets up the controller with the Manager.
+func (r *KongPluginReconciler) SetupWithManager(mgr ctrl.Manager) error {
+	return ctrl.NewControllerManagedBy(mgr).
+		Named("KongPlugin").
+		For(&configurationv1.KongPlugin{}).
+		Watches(
+			&configurationv1alpha1.KongPluginBinding{},
+			handler.EnqueueRequestsFromMapFunc(r.mapKongPluginBindings),
+		).
+		Watches(
+			&configurationv1alpha1.KongService{},
+			handler.EnqueueRequestsFromMapFunc(r.mapKongServices),
+		).
+		Complete(r)
+}
+
+// Reconcile reconciles a KongPlugin object.
+func (r *KongPluginReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	var (
+		entityTypeName = "KongPlugin"
+		logger         = log.GetLogger(ctx, entityTypeName, r.DevelopmentMode)
+	)
+
+	// Fetch the KongPlugin instance
+	var kongPlugin configurationv1.KongPlugin
+	if err := r.Client.Get(ctx, req.NamespacedName, &kongPlugin); err != nil {
+		if k8serrors.IsNotFound(err) {
+			return ctrl.Result{}, nil
+		}
+		return ctrl.Result{}, err
+	}
+
+	log.Debug(logger, "reconciling", kongPlugin)
+
+	// Get the pluginBindings that use this KongPlugin
+	// TODO(mlavacca): use indexers instead of listing all KongPluginBindings
+	pluginBindings := []configurationv1alpha1.KongPluginBinding{}
+	referencingBindingList := configurationv1alpha1.KongPluginBindingList{}
+	err := r.Client.List(ctx, &referencingBindingList, client.InNamespace(kongPlugin.Namespace))
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+	for _, pluginBinding := range referencingBindingList.Items {
+		if pluginBinding.Spec.PluginReference.Name == kongPlugin.Name {
+			pluginBindings = append(pluginBindings, pluginBinding)
+		}
+	}
+
+	// TODO(mlavacca): So far we are supporting only KongService targets here. We need to implement
+	// the same logic for KongRoute, KongConsumer, and KongConsumerGroup as well.
```

Comment:
Can we have an issue for this linked here?

Reply:
sure thing: https://github.com/Kong/gateway-operator/issues/583